### PR TITLE
[CDAP-2258] some refactoring to return better HTTP codes

### PR DESCRIPTION
**Pull request by anew**
Build passed here

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 1]
```diff
@@ -368,11 +369,18 @@ public void performAction(HttpRequest request, HttpResponder responder,
     }
 
     if (!isValidAction(action)) {
-      responder.sendStatus(HttpResponseStatus.METHOD_NOT_ALLOWED);
+      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: I think now we throw exception for error response. E.g. `throw new NotFoundException(...)` in this case. There is a error handler that will turn known exception into corresponding response code and respond.

Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: I was following the existing code. The error handler is - as far as I understand - a fallback for exception thrown by methods further in the call stack. That does not prevent us from explicitly sending a response, does it?

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: No it does not. Looks like we have some mixed style of error handling in this class:

https://github.com/caskdata/cdap/blob/feature/cdap-2258-better-errors/cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java#L231

vs

https://github.com/caskdata/cdap/blob/feature/cdap-2258-better-errors/cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java#L257

**Comment by anew** [comment]
@chtyim I have added a test case.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 1 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: I have tried to keep the refactoring here small. The entire class (and some others - ProgramLifecycleService, DefaultStore) have a wild mix of different error handling styles. This needs more refactoring but that would go beyond a bugfix and I don't want to do this so close to a release.See https://issues.cask.co/browse/CDAP-3107

**Comment by anew** [comment]
@chtyim any other concerns?

**Comment by chtyim** [comment]
LGTM

**Pull request merged**